architecture/spmc.rst

```diff
@@ -0,0 +1,267 @@
+====
+SPMC
+====
+
+This document describes the SPMC (S-EL1) implementation for OP-TEE.
+More information on the SPMC can be found in the FF-A specification can be
+found in the
+`FF-A spec <https://developer.arm.com/documentation/den0077/latest>`_.
+
+.. toctree::
+ :numbered:
+
+SPMC Responsibilities
+=====================
+
+The SPMC is a critical component in the FF-A flow. Some of it mayor
+responsibilities are:
+
+- Initialisation and run-time management of the SPs:
+	The SPMC component is responsible for initialisation of the
+	Secure Partitions (Loading the Image, Setting up the stack, HEAP, ...).
+- Routing messages between endpoints:
+	The SPMC is responsible for passing FF-A messages from The Normal World
+	to SPs and back. It also needs to to pass FF-A messages between SPs.
+- Memory management:
+	The SPMC is responsible for the memory management of the SPs. Memory
+	can be shared between SPs and from a SP to the Normal World.
+
+This document describes OP-TEE as a S-EL1 SPMC.
+
+Secure Partitions
+=================
+Secure Partitions (SPs) are the endpoints used in the FF-A protocol. When
+OP-TEE is used as a SPMC SPs run primarily inside S-EL0.
+
+The SPMC also exposes the OP-TEE core, privileged mode, as an secure endpoint
+itself. This is used to handle all GlobalPlaform programming mode operations
+All GlobalPlatform messages are encapsulated inside FF-A messages.
+The OP-TEE endpoint will unpack the messages and afterwards handle them as
+standard OP-TEE calls. This is needed as TF-A (S-EL3) does only allow
+FF-A messages to be passed to the secure world when the SPMD is enabled.
+
+SPs run from the initial boot of the system until power down and don't have any
+build-in session management compared to user TA. The only means of
+communicating with the outside world is through the FF-A spec defined messages.
+SPs context is saved between executions.
+
+The
+`Trusted Service <https://www.trustedfirmware.org/projects/trusted-services/>`_
+repository includes the libSP libary which export all needed functions to build
+a S-EL0 SP it also include many examples of how to implement SPs. More
```

Comment:
s/include/includes/
add a coma: `...a S-EL0 SP, it also includes...`

Reply:
done